framework/JobHandler.py

```diff
@@ -161,128 +172,117 @@ def __checkAndRemoveFinished(self, running):
+      It initilizes the RAY implementation (with socketing system) in
       case RAVEN is run in a cluster with multiple nodes or the NumMPI > 1,
       otherwise multi-threading is used.
       @ In, None
       @ Out, None
     """
+    ## set up enviroment
+    os.environ['PYTHONPATH'] = os.pathsep.join(sys.path)
     ## Check if the list of unique nodes is present and, in case, initialize the
-    ## socket
+    servers = None
     if self.runInfoDict['internalParallel']:
       if len(self.runInfoDict['Nodes']) > 0:
         availableNodes = [nodeId.strip() for nodeId in self.runInfoDict['Nodes']]
-
-        ## Set the initial port randomly among the user accessible ones
-        ## Is there any problem if we select the same port as something else?
-        randomPort = random.randint(1024,65535)
-
+        ## identify the local host name and get the number of local processors
+        localHostName = self.__getLocalHost()
+        self.raiseADebug("Local host name is  : "+ localHostName)
+        nProcsHead = availableNodes.count(localHostName)
+        self.raiseADebug("# of local procs    : "+ str(nProcsHead))
+        ## initialize ray server with nProcs
+        self.rayServer = ray.init(num_cpus=int(nProcsHead)) if _rayAvail else pp.Server(ncpus=int(nProcsHead))
         ## Get localHost and servers
-        localHostName, ppservers = self.__runRemoteListeningSockets(randomPort)
-        self.raiseADebug("Local host is "+ localHostName)
-
-        if len(ppservers) == 0:
-          ## We are on a single node
-          self.ppserver = pp.Server(ncpus=len(availableNodes))
-        else:
-          ## We are using multiple nodes
-          self.raiseADebug("Servers found are " + ','.join(ppservers))
-          self.raiseADebug("Server port in use is " + str(randomPort))
-          self.ppserver = pp.Server(ncpus=0, ppservers=tuple(ppservers))
+        servers = self.__runRemoteListeningSockets(self.rayServer['redis_address'])
       else:
-         ## We are using the parallel python system
-        self.ppserver = pp.Server(ncpus=int(self.runInfoDict['totalNumCoresUsed']))
+        self.rayServer = ray.init(num_cpus=int(self.runInfoDict['totalNumCoresUsed'])) if _rayAvail else \
+                         pp.Server(ncpus=int(self.runInfoDict['totalNumCoresUsed']))
+      if _rayAvail:
+        self.raiseADebug("Head node IP address: " + self.rayServer['node_ip_address'])
+        self.raiseADebug("Redis address       : " + self.rayServer['redis_address'])
+        self.raiseADebug("Object store address: " + self.rayServer['object_store_address'])
+        self.raiseADebug("Raylet socket name  : " + self.rayServer['raylet_socket_name'])
+        self.raiseADebug("Session directory   : " + self.rayServer['session_dir'])
+        if servers:
+          self.raiseADebug("# of remote servers : " + str(len(servers)))
+          self.raiseADebug("Remote servers      : " + " , ".join(servers))
```

Comment:
No need to change, just FYI, the message handler can handle multiple strings.

Reply:
you are right...much cleaner... i ll do that